File: test-scripts/validate-source.py
#!/usr/bin/env python3
"""
Source-Code Based Astro Migration Validation
Validates the source files before build to ensure:
1. All required components are in place
2. Meta tags and SEO are correct
3. JSON-LD schemas are present
4. WhatsApp integration is complete
5. Form fields are correct
6. Accessibility is maintained
"""

import os
import re
import json
from pathlib import Path
from typing import Dict, List, Any

ROOT_DIR = Path(r"D:\SAMIUDDIN\kilo legend\legend-kilo")
SRC_DIR = ROOT_DIR / "src"

def read_file(filepath: Path) -> str:
    """Read file content."""
    try:
        with open(filepath, 'r', encoding='utf-8') as f:
            return f.read()
    except:
        return ""

def validate_layout() -> Dict[str, Any]:
    """Validate BaseLayout.astro."""
    print("\n" + "="*60)
    print("VALIDATING: src/layouts/BaseLayout.astro")
    print("="*60)
    
    layout_path = SRC_DIR / "layouts" / "BaseLayout.astro"
    content = read_file(layout_path)
    
    results = {
        'exists': layout_path.exists(),
    }
    
    # Check meta tags
    print("\n[Meta Tags]")
    meta_checks = {
        'title': r'<title>{title}</title>',
        'description': r'<meta name="description" content={description}',
        'canonical': r'<link rel="canonical"',
        'og:title': r'<meta property="og:title"',
        'og:description': r'<meta property="og:description"',
        'og:url': r'<meta property="og:url"',
        'robots': r'<meta name="robots"',
    }
    
    for name, pattern in meta_checks.items():
        found = bool(re.search(pattern, content))
        results[f'meta_{name}'] = found
        print(f"  {name}: {'✅' if found else '❌'}")
    
    # Check JSON-LD schemas
    print("\n[JSON-LD Schemas]")
    schemas = re.findall(r'<script type="application/ld\+json">(.*?)</script>', content, re.DOTALL)
    results['schemas_count'] = len(schemas)
    print(f"  Found: {len(schemas)} schemas")
    
    for i, schema in enumerate(schemas):
        try:
            data = json.loads(schema)
            schema_type = data.get('@type', 'Unknown')
            print(f"    {i+1}. {schema_type}")
        except:
            print(f"    {i+1}. (invalid JSON)")
    
    return results

def validate_pages() -> Dict[str, Any]:
    """Validate all page components."""
    print("\n" + "="*60)
    print("VALIDATING: Page Components")
    print("="*60)
    
    pages_dir = SRC_DIR / "pages"
    pages = {
        'index.astro': '/',
        'capabilities.astro': '/capabilities',
        'process.astro': '/process',
        'company.astro': '/company',
        'contact.astro': '/contact',
    }
    
    results = {}
    
    for filename, url_path in pages.items():
        page_path = pages_dir / filename
        content = read_file(page_path)
        
        print(f"\n[{filename}]")
        
        page_results = {
            'exists': page_path.exists(),
        }
        
        # Check page title prop
        title_match = re.search(r'const title = "(.*?)"', content)
        if title_match:
            page_results['title'] = title_match.group(1)
            print(f"  Title: {title_match.group(1)}")
        
        # Check description prop
        desc_match = re.search(r'const description = "(.*?)"', content)
        if desc_match:
            page_results['description'] = desc_match.group(1)[:50] + "..."
            print(f"  Description: {desc_match.group(1)[:50]}...")
        
        # Check for WhatsApp buttons
        wa_count = len(re.findall(r'WhatsAppButton', content))
        page_results['whatsapp_count'] = wa_count
        print(f"  WhatsAppButtons: {wa_count}")
        
        # Check H1 headings
        h1_count = len(re.findall(r'<h1[^>]*id=', content))
        page_results['h1_count'] = h1_count
        print(f"  H1 headings: {h1_count}")
        
        # Check contact page specific
        if filename == 'contact.astro':
            # Check InquiryWidget
            has_inquiry = 'InquiryWidget' in content
            page_results['has_inquiry_widget'] = has_inquiry
            print(f"  InquiryWidget: {'✅' if has_inquiry else '❌'}")
            
            # Check form fields
            form_fields = ['inquiry-company', 'inquiry-product', 'inquiry-quantity', 'inquiry-details']
            for field in form_fields:
                has_field = field in content
                page_results[f'has_{field}'] = has_field
            
            print(f"  Form fields: {sum([page_results.get(f'has_{f}', False) for f in form_fields])}/4")
        
        results[url_path] = page_results
    
    return results

def validate_components() -> Dict[str, Any]:
    """Validate UI components."""
    print("\n" + "="*60)
    print("VALIDATING: UI Components")
    print("="*60)
    
    components_dir = SRC_DIR / "components"
    
    required_components = {
        'Header.astro': 'Header + mobile nav',
        'Footer.astro': 'Footer',
        'WhatsAppButton.astro': 'WhatsApp CTA button',
        'InquiryWidget.astro': 'Inquiry form',
        'SectionHeader.astro': 'Section header',
        'TestimonialCard.astro': 'Testimonial',
        'Icons.astro': 'Icon library',
    }
    
    results = {}
    
    for comp_name, desc in required_components.items():
        comp_path = components_dir / comp_name
        exists = comp_path.exists()
        results[comp_name] = exists
        print(f"  {comp_name}: {'✅' if exists else '❌'} - {desc}")
    
    return results

def validate_data() -> Dict[str, Any]:
    """Validate data layer."""
    print("\n" + "="*60)
    print("VALIDATING: Data Layer")
    print("="*60)
    
    data_path = SRC_DIR / "data" / "site.ts"
    content = read_file(data_path)
    
    results = {
        'exists': data_path.exists(),
    }
    
    # Check key data exports
    exports = [
        'NAV_ITEMS',
        'WHATSAPP_NUMBER',
        'EMAIL_ADDRESS',
        'CAPABILITIES',
        'QUALIFICATION_POINTS',
        'FAQS',
        'PROCESS_STEPS',
    ]
    
    for export in exports:
        found = f'export const {export}' in content
        results[f'has_{export}'] = found
        print(f"  {export}: {'✅' if found else '❌'}")
    
    return results

def validate_accessibility() -> Dict[str, Any]:
    """Validate accessibility features."""
    print("\n" + "="*60)
    print("VALIDATING: Accessibility Features")
    print("="*60)
    
    # Check Header for skip link
    header_path = SRC_DIR / "components" / "Header.astro"
    header_content = read_file(header_path)
    
    results = {}
    
    checks = {
        'skip_link': r'skip to main content',
        'aria_label_nav': r'aria-label="Main navigation"',
        'aria_label_mobile': r'aria-label="Mobile navigation"',
        'aria_expanded': r'aria-expanded',
        'aria_controls': r'aria-controls',
    }
    
    for name, pattern in checks.items():
        found = bool(re.search(pattern, header_content, re.IGNORECASE))
        results[name] = found
        print(f"  Header - {name}: {'✅' if found else '❌'}")
    
    # Check Footer accessibility
    footer_path = SRC_DIR / "components" / "Footer.astro"
    footer_content = read_file(footer_path)
    
    footer_checks = {
        'role_contentinfo': r'role="contentinfo"',
        'aria_label': r'aria-label="Footer',
    }
    
    for name, pattern in footer_checks.items():
        found = bool(re.search(pattern, footer_content, re.IGNORECASE))
        results[f'footer_{name}'] = found
        print(f"  Footer - {name}: {'✅' if found else '❌'}")
    
    return results

def validate_css() -> Dict[str, Any]:
    """Validate Tailwind CSS configuration."""
    print("\n" + "="*60)
    print("VALIDATING: Tailwind CSS Configuration")
    print("="*60)
    
    css_path = SRC_DIR / "styles" / "global.css"
    content = read_file(css_path)
    
    results = {
        'exists': css_path.exists(),
    }
    
    # Check brand colors
    brand_colors = ['brand-50', 'brand-400', 'brand-500', 'brand-600', 'brand-700']
    for color in brand_colors:
        found = color in content
        results[f'has_{color}'] = found
    
    print("  Brand colors:", " ".join([f"{c}{'✅' if results.get(f'has_{c}', False) else '❌'}" for c in brand_colors]))
    
    # Check utility classes
    utilities = ['.btn-primary', '.btn-outline', '.btn-whatsapp', '.card', '.container-site', '.section-padding']
    for util in utilities:
        found = util in content
        results[f'has_{util.replace(".", "")}'] = found
    
    print("  Utilities:", " ".join([f"{u}{'✅' if results.get(f'has_{u.replace(\".\", \"\")}', False) else '❌'}" for u in utilities]))
    
    return results

def validate_config() -> Dict[str, Any]:
    """Validate Astro configuration."""
    print("\n" + "="*60)
    print("VALIDATING: Astro Configuration")
    print("="*60)
    
    config_path = ROOT_DIR / "astro.config.mjs"
    content = read_file(config_path)
    
    results = {
        'exists': config_path.exists(),
    }
    
    checks = {
        'site_url': r"site:\s*'https://legendindustries.in'",
        'tailwind': r'@astrojs/tailwind',
        'output_static': r"output:\s*'static'",
    }
    
    for name, pattern in checks.items():
        found = bool(re.search(pattern, content))
        results[name] = found
        print(f"  {name}: {'✅' if found else '❌'}")
    
    return results

def main():
    print("="*60)
    print("ASTRO MIGRATION - SOURCE CODE VALIDATION")
    print("="*60)
    
    all_results = {}
    
    # Run all validations
    all_results['layout'] = validate_layout()
    all_results['pages'] = validate_pages()
    all_results['components'] = validate_components()
    all_results['data'] = validate_data()
    all_results['accessibility'] = validate_accessibility()
    all_results['css'] = validate_css()
    all_results['config'] = validate_config()
    
    # Summary
    print("\n" + "="*60)
    print("VALIDATION SUMMARY")
    print("="*60)
    
    # Layout
    layout_ok = all_results['layout'].get('meta_title', False) and all_results['layout'].get('schemas_count', 0) >= 3
    print(f"\n[Layout]")
    print(f"  Meta tags: {'✅' if all_results['layout'].get('meta_title', False) else '❌'}")
    print(f"  JSON-LD schemas (3+): {'✅' if all_results['layout'].get('schemas_count', 0) >= 3 else '❌'}")
    
    # Pages
    print(f"\n[Pages]")
    pages_ok = True
    for path, data in all_results['pages'].items():
        wa_count = data.get('whatsapp_count', 0)
        h1_ok = data.get('h1_count', 0) == 1
        print(f"  {path}: WhatsApp={wa_count}, H1={'✅' if h1_ok else '❌'}")
        if wa_count == 0 or not h1_ok:
            pages_ok = False
    
    # Components
    print(f"\n[Components]")
    comps_ok = all(all_results['components'].values())
    print(f"  All required components: {'✅' if comps_ok else '❌'}")
    
    # Data
    print(f"\n[Data Layer]")
    data_ok = all(all_results['data'].values())
    print(f"  All data exports: {'✅' if data_ok else '❌'}")
    
    # Accessibility
    print(f"\n[Accessibility]")
    a11y_ok = all(all_results['accessibility'].values())
    print(f"  All accessibility features: {'✅' if a11y_ok else '❌'}")
    
    # CSS
    print(f"\n[Tailwind CSS]")
    css_ok = all_results['css'].get('exists', False)
    print(f"  Configuration: {'✅' if css_ok else '❌'}")
    
    # Config
    print(f"\n[Astro Config]")
    config_ok = all(all_results['config'].values())
    print(f"  Configuration: {'✅' if config_ok else '❌'}")
    
    # Final verdict
    all_ok = layout_ok and pages_ok and comps_ok and data_ok and a11y_ok and css_ok and config_ok
    
    print("\n" + "="*60)
    if all_ok:
        print("✅ ALL SOURCE VALIDATIONS PASSED")
    else:
        print("⚠️  SOME VALIDATIONS FAILED - See details above")
    print("="*60)
    print("\nNOTE: This validates source code structure only.")
    print("      Full validation requires running the build and testing in browser.")
    print("      Run 'bun install && bun run build && bun run preview' for complete validation.")

if __name__ == "__main__":
    main()

File: test-scripts/validate-source.js
#!/usr/bin/env node
/**
 * Astro Migration - Source Code Validation
 * Validates source files for structure, SEO, accessibility, and integration
 */

const fs = require('fs');
const path = require('path');

const ROOT_DIR = path.join(__dirname, '..');
const SRC_DIR = path.join(ROOT_DIR, 'src');

function readFile(filepath) {
  try {
    return fs.readFileSync(filepath, 'utf8');
  } catch {
    return '';
  }
}

function checkPattern(content, pattern, flags = '') {
  return new RegExp(pattern, flags).test(content);
}

console.log('='.repeat(60));
console.log('ASTRO MIGRATION - SOURCE CODE VALIDATION');
console.log('='.repeat(60));

let allPassed = true;

// 1. Validate Layout
console.log('\n[LAYOUT] src/layouts/BaseLayout.astro');
const layoutPath = path.join(SRC_DIR, 'layouts', 'BaseLayout.astro');
const layoutContent = readFile(layoutPath);

const layoutChecks = {
  'title tag': /<title>\{title\}<\/title>/.test(layoutContent),
  'meta description': /<meta name="description" content=\{description\}/.test(layoutContent),
  'canonical URL': /<link rel="canonical"/.test(layoutContent),
  'og:title': /<meta property="og:title"/.test(layoutContent),
  'og:description': /<meta property="og:description"/.test(layoutContent),
  'og:url': /<meta property="og:url"/.test(layoutContent),
  'robots meta': /<meta name="robots"/.test(layoutContent),
};

const layoutSchemas = (layoutContent.match(/<script type="application\/ld\+json">/g) || []).length;
console.log(`  JSON-LD schemas: ${layoutSchemas}`);

for (const [name, passed] of Object.entries(layoutChecks)) {
  console.log(`  ${name}: ${passed ? '✅' : '❌'}`);
  if (!passed) allPassed = false;
}

// 2. Validate Pages
console.log('\n[PAGES]');
const pages = [
  { file: 'index.astro', path: '/', minWA: 2, hasH1: true },
  { file: 'capabilities.astro', path: '/capabilities', minWA: 1, hasH1: true },
  { file: 'process.astro', path: '/process', minWA: 1, hasH1: true },
  { file: 'company.astro', path: '/company', minWA: 1, hasH1: true },
  { file: 'contact.astro', path: '/contact', minWA: 1, hasH1: true, hasForm: true },
];

for (const page of pages) {
  const pagePath = path.join(SRC_DIR, 'pages', page.file);
  const content = readFile(pagePath);
  
  const waCount = (content.match(/WhatsAppButton/g) || []).length;
  const h1Count = (content.match(/<h1[^>]*id=/g) || []).length;
  const titleMatch = content.match(/const title = "(.*?)"/);
  const descMatch = content.match(/const description = "(.*?)"/);
  
  console.log(`\n  ${page.file}`);
  console.log(`    Title: ${titleMatch ? titleMatch[1].substring(0, 30) + '...' : '❌'}`);
  console.log(`    Description: ${descMatch ? '✅' : '❌'}`);
  console.log(`    WhatsAppButtons: ${waCount} (min: ${page.minWA}) ${waCount >= page.minWA ? '✅' : '❌'}`);
  console.log(`    H1 headings: ${h1Count} ${page.hasH1 ? (h1Count === 1 ? '✅' : '❌') : ''}`);
  
  if (page.hasForm) {
    const hasCompany = content.includes('inquiry-company');
    const hasProduct = content.includes('inquiry-product');
    const hasQuantity = content.includes('inquiry-quantity');
    const hasDetails = content.includes('inquiry-details');
    const hasInquiryWidget = content.includes('InquiryWidget');
    console.log(`    InquiryWidget: ${hasInquiryWidget ? '✅' : '❌'}`);
    console.log(`    Form fields: ${hasCompany && hasProduct && hasQuantity && hasDetails ? '✅ (4/4)' : '❌'}`);
  }
}

// 3. Validate Components
console.log('\n[COMPONENTS]');
const components = [
  'Header.astro',
  'Footer.astro', 
  'WhatsAppButton.astro',
  'InquiryWidget.astro',
  'SectionHeader.astro',
  'TestimonialCard.astro',
  'Icons.astro',
];

for (const comp of components) {
  const compPath = path.join(SRC_DIR, 'components', comp);
  const exists = fs.existsSync(compPath);
  console.log(`  ${comp}: ${exists ? '✅' : '❌'}`);
  if (!exists) allPassed = false;
}

// 4. Validate Data Layer
console.log('\n[DATA LAYER] src/data/site.ts');
const dataPath = path.join(SRC_DIR, 'data', 'site.ts');
const dataContent = readFile(dataPath);

const dataExports = [
  'NAV_ITEMS',
  'WHATSAPP_NUMBER', 
  'EMAIL_ADDRESS',
  'PHONE_NUMBER',
  'CAPABILITIES',
  'QUALIFICATION_POINTS',
  'FAQS',
  'PROCESS_STEPS',
  'BAG_TYPES',
  'CUSTOMIZATION_OPTIONS',
];

for (const exp of dataExports) {
  const found = dataContent.includes(`export const ${exp}`);
  console.log(`  ${exp}: ${found ? '✅' : '❌'}`);
  if (!found) allPassed = false;
}

// 5. Validate Utilities
console.log('\n[UTILITIES]');
const utils = [
  'src/utils/links.ts',
  'src/utils/render-icon.ts',
];

for (const util of utils) {
  const utilPath = path.join(ROOT_DIR, util);
  const exists = fs.existsSync(utilPath);
  console.log(`  ${util}: ${exists ? '✅' : '❌'}`);
  if (!exists) allPassed = false;
}

// 6. Validate Accessibility
console.log('\n[ACCESSIBILITY]');
const headerPath = path.join(SRC_DIR, 'components', 'Header.astro');
const headerContent = readFile(headerPath);

const a11yChecks = {
  'Skip link': /skip to main content/i.test(headerContent),
  'aria-label (nav)': /aria-label="Main navigation"/.test(headerContent),
  'aria-label (mobile)': /aria-label="Mobile navigation"/.test(headerContent),
  'aria-expanded': /aria-expanded/.test(headerContent),
  'aria-controls': /aria-controls/.test(headerContent),
};

for (const [name, passed] of Object.entries(a11yChecks)) {
  console.log(`  Header - ${name}: ${passed ? '✅' : '❌'}`);
  if (!passed) allPassed = false;
}

const footerPath = path.join(SRC_DIR, 'components', 'Footer.astro');
const footerContent = readFile(footerPath);
console.log(`  Footer - role="contentinfo": ${footerContent.includes('role="contentinfo"') ? '✅' : '❌'}`);

// 7. Validate CSS
console.log('\n[TAILWIND CSS] src/styles/global.css');
const cssPath = path.join(SRC_DIR, 'styles', 'global.css');
const cssContent = readFile(cssPath);

const cssChecks = {
  'brand-400': cssContent.includes('brand-400'),
  'brand-500': cssContent.includes('brand-500'),
  'btn-primary': cssContent.includes('.btn-primary'),
  'btn-outline': cssContent.includes('.btn-outline'),
  'btn-whatsapp': cssContent.includes('.btn-whatsapp'),
  'card': cssContent.includes('.card'),
  'container-site': cssContent.includes('.container-site'),
  'section-padding': cssContent.includes('.section-padding'),
};

for (const [name, passed] of Object.entries(cssChecks)) {
  console.log(`  ${name}: ${passed ? '✅' : '❌'}`);
  if (!passed) allPassed = false;
}

// 8. Validate Astro Config
console.log('\n[ASTRO CONFIG] astro.config.mjs');
const configPath = path.join(ROOT_DIR, 'astro.config.mjs');
const configContent = readFile(configPath);

const configChecks = {
  'site URL': /site:\s*'https:\/\/legendindustries\.in'/.test(configContent),
  'Tailwind integration': /@astrojs\/tailwind/.test(configContent),
  'Static output': /output:\s*'static'/.test(configContent),
};

for (const [name, passed] of Object.entries(configChecks)) {
  console.log(`  ${name}: ${passed ? '✅' : '❌'}`);
  if (!passed) allPassed = false;
}

// Summary
console.log('\n' + '='.repeat(60));
if (allPassed) {
  console.log('✅ ALL SOURCE VALIDATIONS PASSED');
} else {
  console.log('⚠️  SOME VALIDATIONS FAILED - See details above');
}
console.log('='.repeat(60));
console.log('\nNOTE: This validates source code structure only.');
console.log('Run "bun install && bun run build" for complete validation.');